server/lib/models/topic.go

```diff
@@ -10,6 +18,147 @@ type Topic struct {
 }
 
 func InitTopic() {
-	create_table(Topic{}, "topics", "Id")
+	t := create_table(Topic{}, "topics", "Id")
+	t.AddIndex("locale", "INDEX", []string{"Locale"})
+}
+
+const TOPIC_FETCH_LIMIT = 50
+
+func NewTopic(dbif Dbif, a *Account, req *proto.TopicCreateRequest) (*Topic, error) {
+	var (
+		p *Persona
+		c *Channel
+		err error
+	)
+	if p, err = a.Persona(dbif, req.Channel); err != nil {
+		return nil, &proto.Err{ Type: proto.Error_TopicCreate_NoPersonaError }
+	}
+	if c, err = FindChannel(dbif, req.Channel); err != nil {
+		return nil, &proto.Err{ Type: proto.Error_TopicCreate_DatabaseError }
+	}
+	body := proto.Model_Topic_Body {
+		ChannelName: c.Name,
+		Name: p.Name,
+	}
+	bs, err := body.Marshal()
+	if err != nil {
+		return nil, &proto.Err{ Type: proto.Error_TopicCreate_DatabaseError }
+	}
+	tp := &Topic {
+		proto.Model_Topic {
+			Id: dbm.UUID(),
+			Channel: req.Channel, 
+			Persona: p.Id,
+			Body: bs,
+		},
+	}
+	if err = dbif.Insert(tp); err != nil {
+		return nil, &proto.Err{ Type: proto.Error_TopicCreate_DatabaseError }
+	}
+	return tp, nil
+}
+
+//return value of actor call
+type HotEntry struct {
+	Id proto.UUID
+	Score int32
+	Vote uint32
+}
+
+func ListTopic(dbif Dbif, req *proto.TopicListRequest) ([]*proto.Model_Topic, error) {
+	var limit int32
+	if req.Limit > 0 {
+		limit = req.Limit
+		if limit > TOPIC_FETCH_LIMIT {
+			limit = TOPIC_FETCH_LIMIT
+		}
+	} else {
+		limit = TOPIC_FETCH_LIMIT
+	}
+	locale := req.Locale
+	if len(locale) <= 0 {
+		locale = "all"
+	}
+	var entries []HotEntry
+	if err := yue.Call(fmt.Sprintf("/hot/%s", locale), "Query", 
+		req.Bucket, req.Query, nil, req.OffsetScore, req.OffsetId, limit, &entries); err != nil {
+		return nil, err
+	}
+	if len(entries) <= 0 {
+		return make([]*proto.Model_Topic, 0), nil
+	}
+	ids := make([]string, len(entries))
+	for i, ent := range entries {
+		ids[i] = strconv.FormatUint(uint64(ent.Id), 10)
+	}
+	tps, err := dbif.Select(Topic{}, dbm.Stmt(`select * from %s.topics where id in (%s)`, strings.Join(ids, ",")))
+	if err != nil {
+		return nil, err
+	}
+	ret := make([]*proto.Model_Topic, len(tps))
+	for i, tp := range tps {
+		tpp := tp.(*Topic).Model_Topic
+		tpp.Point = entries[i].Score
+		tpp.Vote = entries[i].Vote
+		ret[i] = &tpp
+	}
+	return ret, nil
+}
+
+
+
+//insert fixture
+type ChannelInfo struct {
+	Id proto.UUID
+	Name string
 }
 
+func InsertTopicFixture(dbif Dbif) error {
```

Comment:
__[gometalinter]__
* [golint] exported function InsertTopicFixture should have comment or be unexported